### PR TITLE
Feature/#98/active strategy foldout

### PR DESCRIPTION
**Pull request by andre-meyer**
# Description

Adds the active strategy foldout with 3 tabs.

Still missing

- [x] Active Strategies Tab
  - [ ] "Strategies" Tab
    - [ ] Bracket visualization
    - [ ] Deployed parameters
    - [ ] Withdraw button
    - [ ] Balance Tables
  - [x] "Trades" Tab (won't be done in this ticket)
  - [ ] "Deployed params." Tab

# To Test
/

# Background
This PR contains the refactor to overhaul the issues with re-rendering. The balance of tokens has been moved out into a recoil atom (see `state/atoms.ts`)

This PR also changes the default communication method with Infura to websockets, as it has less overhead.

### CODE REVIEW

**Comment by gnosis-info** [comment]
Travis automatic deployment:
 * **🔐 [Safe app link](https://pr127--safecmmappreact.review.gnosisdev.com)**: To be added as a custom app on [Safe Multisig Rinkeby](https://rinkeby.gnosis-safe.io/) or [Mainnet](https://gnosis-safe.io/)
 * **📖 [Storybook](https://pr127--safecmmappreact.review.gnosisdev.com/storybook/)**: Component stories

File: src/hooks/useTokenDetails.ts [hunk 1]
```diff
@@ -1,20 +1,25 @@
...
 interface Return {
-  tokenDetails: TokenBalance | null;
+  tokenDetails?: TokenDetails;
   isLoading: boolean;
   error: string;
 }
 
 export function useTokenDetails(token?: string): Return {
-  const [tokenDetails, setTokenDetails] = useState<TokenBalance | null>(null);
+  const [tokenDetails, setTokenDetails] = useState<TokenDetails>(null);
```
Comment (alfetopito) [inline src/hooks/useTokenDetails.ts:16]: I think this won't work on strict mode, as the `null` value is being set while it now expects a `TokenDetails`

File: src/hooks/useTokenDetails.ts [hunk 2]
```diff
@@ -1,20 +1,25 @@
...
   const [isLoading, setIsLoading] = useState(false);
   const [error, setError] = useState("");
 
-  const { getErc20Details } = useContext(Web3Context);
+  const { getErc20Details } = useContext(
+    ContractInteractionContext
+  ) as ContractInteractionContextProps;
```
Comment (alfetopito) [inline src/hooks/useTokenDetails.ts:22]: Here as well, the casting might no longer be needed

File: src/hooks/useTokenList.ts [hunk 3]
```diff
@@ -10,7 +15,29 @@ import { Web3Context } from "components/Web3Provider";
...
  */
 export function useTokenList(): TokenDetails[] {
-  const { tokenList } = useContext(Web3Context);
+  const [tokenBalances] = useRecoilState(tokenBalancesState);
+  const tokenAddresses = Object.keys(tokenBalances);
+  const [tokenList, setTokenList] = useState([]);
+
+  const { getErc20Details } = useContext(
+    ContractInteractionContext
+  ) as ContractInteractionContextProps;
```
Comment (alfetopito) [inline src/hooks/useTokenList.ts:24]: And here, casting probably not needed

File: src/hooks/useWeb3Strategies.ts [hunk 4]
```diff
@@ -33,15 +42,20 @@ const useWeb3Strategies = (): any => {
...
+    //handleFindStrategies();
+  }, [strategies, handleFindStrategies]);
 
-  useInterval(() => {
+  /*useInterval(() => {
     if (!isFetching) {
       handleFindStrategies();
     }
-  }, 10000);
+  }, 10000);*/
```
Comment (alfetopito) [inline src/hooks/useWeb3Strategies.ts:58]: Clear comments

File: src/hooks/useWeb3Strategies.ts [hunk 5]
```diff
@@ -4,18 +4,27 @@ import useInterval from "@use-it/interval";
...
+
+const useWeb3Strategies = (): Web3StrategyHook => {
   const [status, setStatus] = useState("LOADING");
   const [isFetching, setIsFetching] = useState(false);
   const [strategies, setStrategies] = useState([]);
 
-  const context: Web3ContextType = useContext(Web3Context);
+  const context = useContext(
+    ContractInteractionContext
+  ) as ContractInteractionContextProps;
```
Comment (alfetopito) [inline src/hooks/useWeb3Strategies.ts:27]: Check casting

File: src/logic/strategy.ts [hunk 6]
```diff
@@ -230,10 +240,41 @@ class Strategy {
...
+
+      if (this.quoteTokenAddress && this.baseTokenAddress) {
+        const quoteTokenContract = await context.getContract(
+          "ERC20Detailed",
+          this.quoteTokenAddress
+        );
+        const baseTokenContract = await context.getContract(
+          "ERC20Detailed",
+          this.baseTokenAddress
+        );
```
Comment (alfetopito) [inline src/logic/strategy.ts:252]: Minor: combine into single `Promise.all`

File: src/state/atoms.ts [hunk 7]
```diff
@@ -0,0 +1,3 @@
+import { atom } from "recoil";
+
+export const tokenBalancesState = atom({ key: "tokenBalances", default: {} });
```
Comment (alfetopito) [inline src/state/atoms.ts:3]: Might be helpful to set the shape: `Record<string, BN>`

File: src/types/styled.d.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+// DefaultTheme from styled-components comes empty pre-configured
+// extend it here for correct typings
+declare module "styled-components" {
+  export interface DefaultTheme {
+    colors: {
+      primary: string;
+      secondary: string;
+      white: string;
+      // TODO: Add the rest of the theme variables
```
Comment (alfetopito) [inline src/types/styled.d.ts:11]: Probably import the types from `theme.ts`

File: src/components/basic/inputs/TokenSelector/index.tsx [hunk 9]
```diff
@@ -30,15 +30,13 @@ export const TokenSelector = memo(function TokenSelector(
...
+  const tokenBalance = useTokenBalance(selectedTokenAddress);
 
   // TODO: propagate error to parent component, since the design does not expect errors at the component level
   // probably better when adding validation
-  useEffect((): void => {
-    setError && setError(error);
-  }, [error, setError]);
+  // useEffect((): void => {
+  //   setError && setError(error);
+  // }, [error, setError]);
```
Comment (alfetopito) [inline src/components/basic/inputs/TokenSelector/index.tsx:39]:
```suggestion
```

File: src/components/basic/inputs/Button/index.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+import styled from "styled-components";
+
+import { Button as SRCButton } from "@gnosis.pm/safe-react-components";
+
+// Remove this file when the safe-components team fixes:
+// - color prop only accepts primary, secondary and error when more colors are available
+
+const StyledButton = styled(SRCButton)`
+  .MuiButton-label {
+    color: ${({ color }) => (color === "white" ? "black" : "white")};
```
[the remark this answers is not in the document]
Comment (alfetopito) [reply src/components/basic/inputs/Button/index.js:11]: Never mind, this is about the label color.

**Review by alfetopito — COMMENTED** [review]
Whew, took awhile :)

Several comments, most for myself as I'll be addressing them while you are out.
Also for future reference. 

Feel free to answer them when you are back if you want.

File: src/validators/hasBalance.ts [hunk 11]
```diff
@@ -1,28 +1,30 @@
...
   // Don't validate unless we have all the required input
   if (!tokenAddress || !value) {
     return undefined;
   }
   const details = await getErc20Details(tokenAddress);
   const bnAmount = parseAmount(value, details.decimals);
 
-  if (bnAmount.gt(details.balance)) {
+  const balance = tokenBalances[tokenAddress];
+  if (bnAmount.gt(balance)) {
```
[the remark this answers is not in the document]
Comment (alfetopito) [reply src/validators/hasBalance.ts:27]: 9f4a63c6e3e36b2ebe875bc749944a3aa2b61d2c

File: src/components/pages/strategies/Active/Details.tsx [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+import Strategy from "logic/strategy";
+import { Box, ButtonGroup } from "@material-ui/core";
+import { Button } from "components/basic/inputs/Button";
+
+import { Props as SafeComponentButtonProps } from "@gnosis.pm/safe-react-components/dist/inputs/Button";
+
+// I hate this
+type OldPropsWithoutColor = Omit<SafeComponentButtonProps, "color">;
+type SafeButtonProps = OldPropsWithoutColor & { color: string };
```
[the remark this answers is not in the document]
Comment (alfetopito) [reply src/components/pages/strategies/Active/Details.tsx:11]: I (think I) got why the component is un-stylable.

The new styles are defined within the component https://github.com/gnosis/safe-react-components/blob/development/src/inputs/Button/index.tsx#L26

So styled components cannot override the props and we get type errors even if using the approach I used to override the `size` on the `Text` component https://github.com/gnosis/safe-cmm-app-react/blob/47cb471e5375daf6f2830c2f1a0374e6a3aaf0e2/src/components/basic/display/Text/index.tsx

Alright, I tried to fix it but failed, leaving it as is.

**Review by alfetopito — APPROVED** [review]
Addressed all my comments (self five :hand:)
Fixed a few bugs

Only thing not 100% happy is how the list of tokens is built, based off of token balances global state.
That leaves the initial list of token with only what is on the safe until the balance is fetched for all other tokens, which is about 30s.

Might revisit in a different PR.

Merging now to be able to start merging my changes tomorrow.

**Comment by alfetopito** [comment]
Also, will address whatever is missing from this PR checklist in a separated PR, this is already huge as it stands.

**Comment by andre-meyer** [comment]
> Only thing not 100% happy is how the list of tokens is built, based off of token balances global state.
> That leaves the initial list of token with only what is on the safe until the balance is fetched for all other tokens, which is about 30s.
> 
> Might revisit in a different PR.

Yup! I felt the same way. There needs to be a cooler way to do this, but moving it out of the Context definitely improved things, but now the tokenlist state management sucks. We'll fix this in a different PR for sure. I hated it too 😅 

Thanks so much for your work!

**Pull request merged**